libdnf/rpm/solv/goal_private.cpp

```diff
@@ -632,8 +632,22 @@ transaction::TransactionItemReason GoalPrivate::get_reason(Id id) {
 
     if ((reason == SOLVER_REASON_UNIT_RULE || reason == SOLVER_REASON_RESOLVE_JOB) &&
         (solver_ruleclass(libsolv_solver, info) == SOLVER_RULE_JOB ||
-         solver_ruleclass(libsolv_solver, info) == SOLVER_RULE_BEST))
+         solver_ruleclass(libsolv_solver, info) == SOLVER_RULE_BEST)) {
+        // explicitely user-installed
+        for (const auto & ui_id : user_installed) {
+            if (ui_id == id) {
+                return transaction::TransactionItemReason::USER;
+            }
+        }
+        // explicitely group-installed
+        for (const auto & ui_id : group_installed) {
```

Comment:
It looks like a good place to handle it, but I afraid about side effects. I've been playing with `dnf shell` and I combined remove and install request and I was able to get reinstall operation (it can happen as well with group install and remove). Then the reason for inbound package will be correct but the reason for removed packages will be also group. The installation of group can also trigger upgrade (package installed, new version available with `--best`). It means that replace package will have also reason group. I am really not sure whether it is a problem or not. @lukash Please what is your opinion?

Reply:
(sorry it's a bit of a brain-dump)

I am not sure. First, I'm operating under the assumption that we should record all packages from a group that would be installed by it, but are already installed, as "been installed from the group". That means putting them on the list of installed packages of that group in `groups.toml`.

At the same time they have a reason recorded in `packages.toml`, where I'm not sure if we should use the `GROUP` reason. It would appear we should do it for "consistency" (of the `reason` attribute itself). OTOH, when that group is removed and we should change the `reason` in `packages.toml` to whatever it was before (I guess `EXTERNAL_USER`, `DEPENDENCY` and `WEAK_DEPENDENCY` are the possible values here), I have no idea how to retrieve what was there before without recursing into the history database, which is one of the core design concepts that we don't do that.

I don't even know what should be the reason for the `REMOVE` action, because there the semantics shift (the `reason` becomes the reason of why the package was removed, as opposed to why the package is installed). I guess for printing out the transaction table having `GROUP` there is good?

For upgrades etc., I guess the reason `GROUP` is good, as we do want to mark the package as installed by the group. But then, we need that for even packages that aren't even being upgraded (they are just already installed).

And, on top of all that, we don't only need the `reason` for the packages installed from a group, we also need the names of groups that they were installed by, so that we can record it in both system state and the history db.